lib/record_store/zone.rb

```diff
@@ -189,7 +189,9 @@ def build_authority(authority)
     end
 
     def build_records(records)
-      records.map { |record| Record.build_from_yaml_definition(record) }
+      all_records = records.map { |record| Record.build_from_yaml_definition(record) }
+      config.implicit_record_templates.each { |template| template.inject_records!(all_records: all_records) }
```

Comment:
It feels weird that you're modifying the original records here. Consider passing them in and having the template return the modified records

Reply:
we want to modify the original records so that each iteration of template injection does its work with all the records in the zone up to that point (i.e. with full context). otherwise we have to do a tonne of conflict resolution in this method to make sure all the records returned by the templates dont conflict with each other

Reply:
Why not something like
```
records = all_records
config.implicit_record_templates.each do |template|
  records = template.inject_records!(records)
end
```

Reply:
hmm ruby is pass by reference so that ^ does pretty much the same as whats here. would it be better if we create an empty array thats a duplicate of `all_records`, pass that array around between templates, and then return the cumulative array at the end, vs modifying the original?

Reply:
Since the original is generated using `map`, it's already a "new array" though.
What @dschwar offers is nice to read, but to have a method that works that way would, as you said, require to either `.dup` the array at the beginning, or return just the records to inject and merge them here.

That could be something like:
```suggestion
      config.implicit_record_templates.each { |template| all_records.push(*template.records_to_inject(current_records: all_records)) }
```

With `records_to_inject` returning either an empty list (nothing to inject) or a list containing new records to inject. That would also keep the `all_records` list up to date. What are your thoughts on this ?

Also, nitpick: I renamed the parameter as `current_records`. I feel that in the `record_store` we generally avoid named parameters unless they're necessary, but I don't mind much having an extra name here, I suggest however to name it `current_records` instead of `all_records` as it feels clearer about what it means. ("all records" is for me the final version, with everything injected, in some way)

Reply:
i don't have strong feelings here - if we feel its clearer to do it this way, I'm happy to make the change. this way still lets us delegate conflict resolution to the individual templates with all relevant context provided, which is what my main requirement would be, so i'm good with it.

Reply:
No strong feeling either. I'd say, since the change for this is minimal, and it might make things slightly clearer when reading (making it directly seen that we're changing the `all_records` list, instead of hidden in the method call), let's go for it!